### PR TITLE
Add code value trimming "behaviors" for compatibility with old iModels, and minor error handling cleanup

### PR DESCRIPTION
**Pull request by MichaelBelousov**
In #180 we started unicode-aware (previously ascii) trimming of code values. Before that, several production iModels stored code values containing non-ascii whitespace. A couple transformer library users that are trying to move up to 4.x have reported failures when transforming these iModels since the codes are not preserved during db loads/stores.

Since the transformer wants to preserve the code values as they are stored in the Db, this PR introduces a per-dgndb `DgnCodeValue::Behavior m_codeValueBehavior`, and makes initialization of any DgnCodeValue tied to a dgndb context, e.g. FromJson, LoadFromDb. The default behavior is `TrimUnicodeWhiteSpace`, but now you can also set it to `Exact`, exposed to javascript as `nativeDb.get/setCodeValueBehavior(val: "exact" | "trim-unicode-whitespace")`. The intention is for users like the transformer to change the handling during a transformation, and then turn it back.

itwinjs-core: https://github.com/iTwin/itwinjs-core/pull/5861

This PR also prevents a nullptr dereference segfault and instead throws descriptively in some cloning code that is reached if a transformation occurs when schemas haven't been exported between two iModels. 

Misc:
- I intend to backport this as a patch to the current release after this merges to master

### CODE REVIEW

**Review by ColinKerr — APPROVED** [review]

File: iModelCore/iModelPlatform/Tests/DgnProject/NonPublished/CodeSpec_Test.cpp [hunk 1]
```diff
@@ -121,20 +121,36 @@ TEST_F(DgnCodeSpecsTest, ImportCodeSpec)
...
+    ASSERT_STREQ(value, DgnCodeValue("Value\n", DgnCodeValue::Behavior::TrimUnicodeWhitespace).GetUtf8CP());
+    ASSERT_STREQ(value, DgnCodeValue("\nValue\n", DgnCodeValue::Behavior::TrimUnicodeWhitespace).GetUtf8CP());
+    ASSERT_STREQ(value, DgnCodeValue("  \t\nValue \t \n  ", DgnCodeValue::Behavior::TrimUnicodeWhitespace).GetUtf8CP());
     ASSERT_TRUE(DgnCodeValue().empty());
-    ASSERT_TRUE(DgnCodeValue(" ").empty());
-    ASSERT_TRUE(DgnCodeValue(" \t\n ").empty());
+    ASSERT_TRUE(DgnCodeValue(" ", DgnCodeValue::Behavior::TrimUnicodeWhitespace).empty());
+    ASSERT_TRUE(DgnCodeValue(" \t\n ", DgnCodeValue::Behavior::TrimUnicodeWhitespace).empty());
+
+    #define ASSERT_EXACT(v) ASSERT_STREQ((v), DgnCodeValue((v), DgnCodeValue::Behavior::Exact).GetUtf8CP())
```
Comment (nick4598) [inline iModelCore/iModelPlatform/Tests/DgnProject/NonPublished/CodeSpec_Test.cpp:139]: This part of the tests looks wrong to me, there are back to back ASSERT_EXACTS for multiple lines

Comment (MichaelBelousov) [reply iModelCore/iModelPlatform/Tests/DgnProject/NonPublished/CodeSpec_Test.cpp:139]: yep realized, I glossed over the result of my vim macro like an idiot... rerunning fixed tests locally and will push once it succeeds

Comment (MichaelBelousov) [reply iModelCore/iModelPlatform/Tests/DgnProject/NonPublished/CodeSpec_Test.cpp:140]: fixed

**Review by nick4598 — APPROVED** [review]

**Review by khanaffan — APPROVED** [review]

**Review by aruniverse — APPROVED** [review]

**Comment by aruniverse** [comment]
@mergifyio backport release/4.x

**Comment by MichaelBelousov** [comment]
waiting on #5862

**Review by pmconne — COMMENTED** [review]
Are you expecting to have to expand the `Behavior` enum in future?
This would be a much smaller and clearer change if `DgnCodeValue` constructor took a `bool trim=true`, or if you replaced `Behavior { TrimUnicodeWhitespace, Exact }` with `Trim { No, Yes }` with `Yes` being the default.

**Comment by pmconne** [comment]
If you're going to make the `Behavior` argument required, please document that enum member to make it crystal clear it should be the default choice.

**Comment by MichaelBelousov** [comment]
> If you're going to make the `Behavior` argument required, please document that enum member to make it crystal clear it should be the default choice.

done

**Review by pmconne — APPROVED** [review]

**Comment by MichaelBelousov** [comment]
OOM for [android build](https://dev.azure.com/bentleycs/iModelTechnologies/_build/results?buildId=2337888&view=logs&j=023ba795-95cf-53ec-173c-c0bcaae2690b&t=712cbd68-614e-5470-00b0-5caeb3baa3b2&s=25d9f337-c615-5e4f-e36c-693c7c0ae31a)?:

```

~putenv DEBUG=
~putenv GRADLE_OPTS=-Xms4096m -Xmx4096m

D:\vsts_a\8\b\Tools\gradle\bin\gradle --project-dir D:\vsts_a\8\b\AndroidARM64\DgnPlatTst\prj --no-daemon --quiet build assembleAndroidTest
D:/vsts_a/8/b/Tools/androidndk_r25c/toolchains/llvm/prebuilt/windows-x86_64/bin/llvm-strip.exe: error: 'D:\vsts_a\8\b\AndroidARM64\DgnPlatTst\prj\app\build\intermediates\merged_native_libs\debug\out\lib\arm64-v8a\gdb.setup': The file was not recognized as a valid object file

D:/vsts_a/8/b/Tools/androidndk_r25c/toolchains/llvm/prebuilt/windows-x86_64/bin/llvm-strip.exe: error: 'D:\vsts_a\8\b\AndroidARM64\DgnPlatTst\prj\app\build\intermediates\merged_native_libs\release\out\lib\arm64-v8a\gdb.setup': The file was not recognized as a valid object file

[37.298s][warning][gc,alloc] WorkerExecutor Queue Thread 27: Retried waiting for GCLocker too often allocating 6470945 words

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':app:mergeReleaseNativeDebugMetadata'.
> A failure occurred while executing com.android.build.gradle.internal.tasks.MergeNativeDebugMetadataTask$MergeNativeDebugMetadataWorkAction
   > Java heap space
```

**Comment by pmconne** [comment]
> OOM for android build

It happens. Rerun it. @tcobbs-bentley @swbsi

**Comment by MichaelBelousov** [comment]
third attempt at android build

**Comment by swbsi** [comment]
> third attempt at android build

Increasing the -Xms value in the gradle options should fix this.

**Comment by MichaelBelousov** [comment]
> > third attempt at android build
> 
> Increasing the -Xms value in the gradle options should fix this.

if it happens again I can add that to this PR

**Pull request merged**